File: src/2015/Day_03/P2.rb
data = File.read("input.txt").split("")

houses = ["0:0"]
checked = 1

posS = [0, 0]
posR = [0, 0]

i = 0
for token in data
    ch = i % 2 === 0

    if token === "<"
        ch ? posS[0] -= 1 : posR[0] -= 1
    elsif token === ">"
        ch ? posS[0] += 1 : posR[0] += 1
    elsif token === "^"
        ch ? posS[1] += 1 : posR[1] += 1
    else
        ch ? posS[1] -= 1 : posR[1] -= 1
    end

    coord = "#{ch ? posS[0] : posR[0]}:#{ch ? posS[1] : posR[1]}"
    i += 1

    if houses.include? coord
        next
    else
        houses.append(coord)
        checked += 1
    end
end

puts "Result: #{checked}"

File: src/2022/Day_03/P2.js
import * as path from 'path';
import * as fs from 'fs';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const data = fs.readFileSync(path.resolve(__dirname, `./input.txt`), `utf-8`);

const priority = `abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ`.split(``);

const main = async () => {
    const lines = data.split(`\n`).map(x => x.trim());
    let ans = 0;

    for (let i = 0; i < lines.length; i += 3) {
        const group = lines.slice(i, i + 3);

        let commonToken;
        for (const token of group[0]) {
            if (group.every(x => x.includes(token))) {
                commonToken = token;
                break;
            }
        }

        ans += priority.indexOf(commonToken) + 1;
    }

    fs.writeFileSync(path.resolve(__dirname, `./output.txt`), String(ans));

    if (process.argv.length <= 3) console.log(`Result: ${ans}`);
    return ans;
};

if (process.argv.length <= 3) void main();
export default main;
